src/mlpack/core/dists/diagonal_gaussian_distribution.cpp

```diff
@@ -0,0 +1,155 @@
+{
+  this->covariance = covariance;
+  InvertCovariance();
+  LogDeterminant();
+}
+
+void DiagonalGaussianDistribution::Covariance(arma::vec&& covariance)
+{
+  this->covariance = std::move(covariance);
+  InvertCovariance();
+  LogDeterminant();
+}
+
+double DiagonalGaussianDistribution::LogProbability(
+    const arma::vec& observation) const
+{
+  const size_t k = observation.n_elem;
+  const arma::vec diff = observation - mean;
+  const arma::vec logExponent = (diff.t() * arma::diagmat(invCov) * diff);
+  return -0.5 * k * log2pi - 0.5 * logDetCov - 0.5 * logExponent(0);
+}
+
+void DiagonalGaussianDistribution::InvertCovariance()
+{
+  // Calculate the inverse of the diagonal covariance.
+  invCov = 1/covariance;
+}
+
+void DiagonalGaussianDistribution::LogDeterminant()
+{
+  // Calculate Log determinant of the diagonal covariance.
+  logDetCov = arma::accu(log(covariance));
+}
+
+arma::vec DiagonalGaussianDistribution::Random() const
+{
+  return (arma::sqrt(covariance) % arma::randn<arma::vec>(mean.n_elem)) + mean;
+}
+
+void DiagonalGaussianDistribution::Train(const arma::mat& observations)
+{
+  if (observations.n_cols > 1)
+  {
+    covariance.zeros(observations.n_rows);
+  }
+  else
+  {
+    mean.zeros(0);
+    covariance.zeros(0);
+    return;
+  }
+
+  // Calculate the mean.
+  mean = arma::sum(observations, 1);
+
+  // Normalize the mean.
+  mean /= observations.n_cols;
+
+  // Now calculate the covariance.
+  const arma::mat diffs = observations - mean *
+      arma::ones<arma::rowvec>(observations.n_cols);
+  covariance += arma::sum(diffs % diffs, 1);
+
+  // Finish estimating the covariance by normalizing, with the (1 / (n - 1))
+  // to make the estimator unbiased.
+  covariance /= (observations.n_cols - 1);
+
+  InvertCovariance();
+  LogDeterminant();
+}
+
+void DiagonalGaussianDistribution::Train(const arma::mat& observations,
+                                         const arma::vec& probabilities)
+{
+  if (observations.n_cols > 0)
+  {
+    covariance.zeros(observations.n_rows);
+  }
+  else
+  {
+    mean.zeros(0);
+    covariance.zeros(0);
+    return;
+  }
+
+  // We'll normalize the covariance with (v1 - (v2 / v1))
+  // for unbiased estimator in the weighted arithmetic mean. The v1 is the sum
+  // of the weights, and the v2 is the sum of the each weight squared.
+  // If you want to know more detailed description,
+  // please refer to https://en.wikipedia.org/wiki/Weighted_arithmetic_mean
+  double v1 = arma::accu(probabilities);
+  arma::vec normalizedProbs = probabilities;
+
+  // If their sum is 0, there is nothing in this Gaussian.
+  // At least, set the covariance so that it's invertible.
+  if (v1 == 0)
+  {
+    covariance += 1e-50;
+    InvertCovariance();
+    LogDeterminant();
+    return;
+  }
+
+  // If their sum is not 1, divide the weights by them to normalize.
+  if (v1 != 1)
+  {
+    normalizedProbs = probabilities / v1;
+    v1 = arma::accu(normalizedProbs);
+  }
```

Comment:
Actually reading through all this code, it would be faster to do this:

```
double v1 = arma::accu(probabilities);
if (v1 == 0)
{
  covariance += 1e-50;
  InvertCovariance();
  LogDeterminant();
  return;
}

arma::vec normalizedProbs = probabilities / v1;
```

(This avoids a copy of `probabilities` and a call to `arma::accu(normalizedProbs)`.)

Reply:
Thanks for letting me know!! :+1: